src/app_client.rs

```diff
@@ -38,6 +41,169 @@ use crate::{
 	types::{AppClientConfig, BlockVerified},
 };
 
+#[async_trait]
+#[automock]
+pub trait AppClient {
+	async fn reconstruct_rows_from_dht(
+		&self,
+		pp: PublicParameters,
+		block_number: u32,
+		dimensions: &Dimensions,
+		commitments: &[[u8; config::COMMITMENT_SIZE]],
+		missing_rows: &[u32],
+	) -> Result<Vec<(u32, Vec<u8>)>>;
+
+	async fn fetch_rows_from_dht(
+		&self,
+		block_number: u32,
+		dimensions: &Dimensions,
+		row_indexes: &[u32],
+	) -> Vec<Option<Vec<u8>>>;
+
+	async fn get_kate_rows(&self, rows: Vec<u32>, block_hash: H256)
+		-> Result<Vec<Option<Vec<u8>>>>;
+
+	fn store_encoded_data_in_db<T: Encode + 'static>(
+		&self,
+		app_id: u32,
+		block_number: u32,
+		data: &T,
+	) -> Result<()>;
+}
+
+#[derive(Clone)]
+pub struct AppClientImpl {
+	db: Arc<DB>,
+	network_client: Client,
+	rpc_client: OnlineClient<AvailConfig>,
+}
+
+#[async_trait]
+impl AppClient for AppClientImpl {
+	async fn reconstruct_rows_from_dht(
```

Comment:
This fn will be untested since its mocked, so in some next iteration we should think how to test it.